File: challenge1/Java/CountSequences.java
import java.io.*;
import java.util.*;
import java.util.stream.*;
import static java.util.stream.Collectors.toList;

public class Solution {
    
    private static int countSequences(int idx, int currSum, int targetSum, List<List<Integer>> numRanges){
        if (idx >= numRanges.size()){
            if (currSum == targetSum){
                return 1;
            }
            
            return 0;
        }
        
        int l = numRanges.get(idx).get(0);
        int r = numRanges.get(idx).get(1);
        
        int cnt = 0;
        for(int i = l; i <= r; i++){
            // if currSum + i > targetSum then don't bother
            cnt += countSequences(idx + 1, currSum + i, targetSum, numRanges);
        }
        
        return cnt;
    }

    public static void main(String[] args) throws IOException {
        /* Enter your code here. Read input from STDIN. Print output to STDOUT. Your class should be named Solution. */
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        // n and S
        String[] firstLineInput = bufferedReader.readLine().replaceAll("\\s+$", "").split(" ");
        int n = Integer.parseInt(firstLineInput[0]);
        int s = Integer.parseInt(firstLineInput[1]);
        
        List<List<Integer>> numRanges = new ArrayList<>();
        IntStream.range(0, n).forEach(i -> {
            try {
                numRanges.add(
                    Stream.of(bufferedReader.readLine().replaceAll("\\s+$", "").split(" "))
                        .map(Integer::parseInt)
                        .collect(toList())
                );
            } catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        });
        
        int result = countSequences(0, 0, s, numRanges);
        
        bufferedWriter.write(String.valueOf(result));
        bufferedWriter.newLine();

        bufferedReader.close();
        bufferedWriter.close();
    }
}

File: challenge1/Javascript/countSequences.js
function processData(input) {
    let args = input.split("\n");
    let iterator = args[Symbol.iterator]();
    let numRange = []

    let current = iterator.next();
    let target = parseInt(current.value.split(" ")[1]);

    while (!current.done){
        current = iterator.next();
        let value = current.value;
        if (value){
            let array = [];
            value.split(" ").forEach(number=> array.push(parseInt(number)));
            numRange.push(array);
        }
    }
    let result = countSequences(0, 0, target, numRange);
    console.log(result);
}

function countSequences(index, currSum, target, numRange){
    if (index >= numRange.length){
        if (currSum === target){
            return 1;
        }
        return 0;
    }

    let l = numRange[index][0];
    let r = numRange[index][1];
    let count = 0;

    for(let i = l; i <= r; i++){
        count += countSequences(index+1, currSum+i, target, numRange);
    }
    return count;
}

process.stdin.resume();
process.stdin.setEncoding("ascii");
_input = "";
process.stdin.on("data", function (input) {
    _input += input;
});

process.stdin.on("end", function () {
    processData(_input);
});